tensorboard/uploader/proto/write_service.proto

```diff
@@ -119,6 +121,82 @@ message WriteScalarResponse {
   // This is empty on purpose.
 }
 
+// The blob writing procedure works as follows:
+//  * The caller instantiates a WriteBlob RPC, which is a bidirectional stream.
+//  * The caller sends a number of WriteBlobRequests, providing the blob data
+//    in an ordered sequence of chunks.
+//  * The request size should be less than 4 MiB (2^20 = 4,194,304), because
+//    that is the default request limit for gRPC.  Accounting for the other
+//    fields of the request, then, and leaving some padding for safety, the data
+//    chunk should be somewhat smaller-- perhaps simply 4e6 bytes.
+//  * The first of these requests must provide write_blob_first_request_data,
+//    and the subsequent ones must not.
+//  * The last of these requests must set finalize_object=True, and the prior
+//    ones must not.
+//  * For a blob <= 4MB, a single chunk suffices, so the single request is both
+//    the first and the last.
+//  * The caller receives a sequence of WriteBlobResponses in return.  Each
+//    WriteBlobResponse corresponds to one WriteBlobRequest in order, reporting
+//    the state of the write operation.
+//  * The caller may choose to verify each response before sending the next
+//    request, or it may continue sending requests optimistically.
+//
+// Note that the WriteBlobRequest does not mirror the nested structure of
+// WriteScalarRequest, because we only ever send one blob at a time.
+
+// Metadata included with only the first chunk of the blob write stream.
+message WriteBlobFirstRequestData {
+  string experiment_id = 1;
+  string run = 2;
+  string tag = 3;
+  int64 step = 4;
+  int64 index = 5;
+  // Note that metadata.plugin_data.content does not carry the payload.
+  .tensorboard.SummaryMetadata metadata = 6;
+}
+
+// A single chunk of the blob write stream.
+message WriteBlobRequest {
+  WriteBlobFirstRequestData write_blob_first_request_data = 1;
+  bytes data = 2 [ctype = CORD];
+  // The position in the blob where this chunk begins.
+  // This must equal the sum of the sizes of the chunks sent so far.
+  int64 offset = 3;
+  // CRC32C of current data buffer. While optional it is recommended that
+  // clients include the crc32c for every data buffer, to protect against data
+  // corruption. Note that for multi-shot writes, specifying the crc32c for
+  // every data buffer provides stronger protection than just providing
+  // the final_crc32c at the end of the upload.
+  fixed32 crc32c = 4;
+  // Indicates that this is the last chunk of the stream.
+  bool finalize_object = 5;
+  // CRC32C of the entire blob. Optional but recommended, to protect against
+  // data corruption. This should be set only when finalize_object=True.
+  fixed32 final_crc32 = 6;
+}
+
+message WriteBlobResponse {
+  enum ObjectState {
```

Comment:
I don’t understand what this `ObjectState` means or how the client will
use it. If the client sends the final chunk of a blob and the server
responds `UNFINALIZED` because the blob is still being asynchronously
written to the backing store, there’s no way that the client will be
able to see `CURRENT`, right, because there are no more request–response
pairs to be sent?

Reply:
The server shouldn't send any response at all until the backing write is complete.  That's how things should work whether the server is using sync or async RPCs to backing storage.  In the sync case it's clear: the response to the client is placed on the stream after the backing write succeeds.  In the async case, the response to the client is placed on the stream in the backing RPC callback-- again, after the write succeeds.

A key point here is that the client RPC, as a bidirectional stream, has a sort-of-async quality: the client may choose to ignore the response stream while continuing to send requests, but then still look at the responses later.  At least, that's what I *thought* would happen, but now it appears that's [not directly supported yet](https://github.com/grpc/grpc/issues/18376).  The default behavior is basically synchronous after all (i.e., the client receives each response before sending the next request).  There is a (kinda messy) [workaround](https://stackoverflow.com/questions/55029342/handling-async-streaming-request-in-grpc-python) though.

So I think my (strawman) plan is:
 * Phase 1: sync calls everywhere, for simplicity.
 * Phase 2: async calls from the server to storage.  This is not too hard, and means that the client will still block but the server won't.
 * Phase 3: Apply the async workaround above to the client.

The latter two are optimizations re resource usage and bandwidth, so I think they can be postponed for a short time at least.

Reply:
Discussed offline; sounds like the above plan is ok.